commit 3f6bae37667e55cdf243909dd69990ab27ce4a15
Author: kikairoya <kikairoya@gmail.com>
Date:   Sun Sep 14 18:58:58 2014 +0900

    Revert "eliminate install of duplicate headers (take 2)"
    
    This reverts commit 44ead612371ad529cf06d0a454a4df52738edae4.

diff --git a/CMakeLists.txt b/CMakeLists.txt
index c20e5d1..dd36b61 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -90,17 +90,16 @@ set(LIBCXX_TARGET_TRIPLE ${LIBCXX_TARGET_TRIPLE} CACHE STRING "Target triple.")
 macro(setup_abi_lib abipathvar abidefines abilibs abifiles abidirs)
   list(APPEND LIBCXX_CXX_FEATURE_FLAGS ${abidefines})
   set(${abipathvar} "${${abipathvar}}"
-    CACHE PATH
-    "Paths to C++ ABI header directories separated by ';'." FORCE
+    CACHE STRINGS
+    "Paths to ABI include directories separate by ';'."
     )
   set(LIBCXX_CXX_ABI_LIBRARIES ${abilibs})
   set(LIBCXX_ABILIB_FILES ${abifiles})
-
   file(MAKE_DIRECTORY "${CMAKE_BINARY_DIR}/include")
   foreach(_d ${abidirs})
     file(MAKE_DIRECTORY "${CMAKE_BINARY_DIR}/include/${_d}")
   endforeach()
-
+  set(LIBCXX_ABILIB_FILE_PATHS)
   foreach(fpath ${LIBCXX_ABILIB_FILES})
     set(found FALSE)
     foreach(incpath ${${abipathvar}})
@@ -108,23 +107,29 @@ macro(setup_abi_lib abipathvar abidefines abilibs abifiles abidirs)
         set(found TRUE)
         get_filename_component(dstdir ${fpath} PATH)
         get_filename_component(ifile ${fpath} NAME)
-        file(COPY "${incpath}/${fpath}"
-          DESTINATION "${CMAKE_BINARY_DIR}/include/${dstdir}"
+        add_custom_command(
+          OUTPUT "${CMAKE_BINARY_DIR}/include/${dstdir}/${ifile}"
+          COMMAND ${CMAKE_COMMAND} -E copy_if_different
+            "${incpath}/${fpath}"
+            "${CMAKE_BINARY_DIR}/include/${dstdir}"
+          MAIN_DEPENDENCY "${incpath}/${fpath}"
+          )
+        list(APPEND LIBCXX_CXX_ABI_DEPS
+          "${CMAKE_BINARY_DIR}/include/${dstdir}/${ifile}"
           )
-        list(APPEND abilib_headers "${CMAKE_BINARY_DIR}/include/${fpath}")
       endif()
     endforeach()
     if (NOT found)
       message(FATAL_ERROR "Failed to find ${fpath}")
     endif()
   endforeach()
-
-  add_custom_target(LIBCXX_CXX_ABI_DEPS DEPENDS ${abilib_headers})
+  add_custom_target(abilib_headers DEPENDS ${LIBCXX_CXX_ABI_DEPS})
+  set(LIBCXX_CXX_ABI_DEPS abilib_headers)
   include_directories("${CMAKE_BINARY_DIR}/include")
-
-  install(FILES ${abilib_headers}
+  install(DIRECTORY "${CMAKE_BINARY_DIR}/include/"
     DESTINATION include/c++/v1
-    PERMISSIONS OWNER_READ OWNER_WRITE GROUP_READ WORLD_READ
+    FILES_MATCHING
+    PATTERN "*"
     )
 endmacro()
 
@@ -176,7 +181,6 @@ if (MSVC)
 else()
   if (LIBCXX_HAS_NOSTDINCXX_FLAG)
     list(APPEND LIBCXX_CXX_REQUIRED_FLAGS -nostdinc++)
-    string(REPLACE "-stdlib=libc++" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
   endif()
   if (LIBCXX_ENABLE_CXX0X AND LIBCXX_HAS_STDCXX0X_FLAG)
     list(APPEND LIBCXX_CXX_REQUIRED_FLAGS -std=c++0x)
diff --git a/include/CMakeLists.txt b/include/CMakeLists.txt
index bbf7ea4..0c88b7f 100644
--- a/include/CMakeLists.txt
+++ b/include/CMakeLists.txt
@@ -2,12 +2,15 @@ if (NOT LIBCXX_INSTALL_SUPPORT_HEADERS)
   set(LIBCXX_SUPPORT_HEADER_PATTERN PATTERN "support" EXCLUDE)
 endif()
 
-install(DIRECTORY .
-  DESTINATION include/c++/v1
+file(COPY .
+  DESTINATION "${CMAKE_BINARY_DIR}/include/c++/v1"
   FILES_MATCHING
   PATTERN "*"
   PATTERN "CMakeLists.txt" EXCLUDE
   PATTERN ".svn" EXCLUDE
   ${LIBCXX_SUPPORT_HEADER_PATTERN}
-  PERMISSIONS OWNER_READ OWNER_WRITE GROUP_READ WORLD_READ
+  )
+
+install(DIRECTORY "${CMAKE_BINARY_DIR}/include/c++/v1/"
+  DESTINATION include/c++/v1/
   )
diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index 70adcb7..cec0bee 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -32,7 +32,7 @@ else()
 endif()
 
 if (DEFINED LIBCXX_CXX_ABI_DEPS)
-  add_dependencies(cxx LIBCXX_CXX_ABI_DEPS)
+  add_dependencies(cxx ${LIBCXX_CXX_ABI_DEPS})
 endif()
 
 # Generate library list.
